src/main/kotlin/com/geekbeast/rhizome/jobs/AbstractDistributedJob.kt

```diff
@@ -0,0 +1,260 @@
+            this.status = JobStatus.CANCELED
+        }
+        publishJobState()
+        logger.error(msg())
+    }
+
+    fun pause(msg: () -> String = { "Job $id with $taskId was aborted." }) {
+        if (this.status != JobStatus.FINISHED) {
+            this.status = JobStatus.PAUSED
+        }
+        publishJobState()
+        logger.info(msg())
+    }
+
+    abstract fun processNextBatch()
+
+    private fun initializeTask() {
+        /**
+         * There are two resumption cases we need to handle at initialization
+         *
+         * 1. Node failure and task is resumed on another node and task only has original state
+         * 2. Cluster restart and task is being resumed and task has last previously published state.
+         *
+         * In the first case we have to pull state from map to resume the job. In the second case we
+         * will already have the state from map and just need to resume the job.
+         */
+
+        status = jobs.executeOnKey(id!!) { it.value.status }
+
+        if (resumable && status == JobStatus.RUNNING) {
+            //Case 1
+            val v = jobs.executeOnKey(id!!) { it.value }!!
+            this.state = state
+            this.status = v.status
+            this.taskId = v.taskId
+            this.hasWorkRemaining = v.hasWorkRemaining
+            this.progress = progress
+        } else if (status == JobStatus.PENDING) {
+            //If status is pending then job is new and we just need to mark as running.
+            status = JobStatus.RUNNING
+        }
+    }
+
+    private fun processBatches() {
+        while (!interrupted() && hasWorkRemaining && status == JobStatus.RUNNING) {
+            processNextBatch()
+            if (!hasWorkRemaining) {
+                status = JobStatus.FINISHED
+            }
+            publishJobState()
+        }
+    }
+
+    final override fun call(): R? {
+        //Wait until task id has been assigned and can be retrieved.
+        initializeTaskId()
+
+        /**
+         * If the job status is paused we poll for a job status update every five seconds, otherwise we process
+         * that batches, which publishes it's state on every loop. If we get interrupted, we abort the processing
+         * and publish the aborted job state.
+         */
+        try {
+            do {
+                if (status == JobStatus.PAUSED) {
+                    Thread.sleep(PAUSE_POLLING_MILLIS)
+                }
+                initializeTask()
+                processBatches()
+            } while (resumable && status == JobStatus.PAUSED)
+        } catch (ex: InterruptedException) {
+            logger.warn("Distributed job $taskId was interrupted.", ex)
+            abort()
+        } finally {
+            publishJobState()
+        }
+
+        return if (hasWorkRemaining && status == JobStatus.CANCELED) {
+            logger.warn("Distributed job $id with task id $taskId was interrupted or canceled.")
```

Comment:
Also minor, but I like the practice of using the {} syntax to avoid potentially unnecessary string interpolation in these messages depending on the log level. like `logger.warn("Distributed job {} with task id {} was interrupted or canceled.", id, taskId)` <-- that syntax, whatever it's called